### PR TITLE
Fix Blazor TextArea Not Updating

### LINKED ISSUE #45951: Blazor TextArea Not Updating
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

If you have a <textarea>, which needs to get its data from a method, not a bindable property, and we also want to save any changes made by the user to the text in that <textarea>, one may think to write the following code:
```
<textarea rows="3" style="width: 350px;" @onchange="Text_OnChange">@GetData(someParameter)</textarea>
```
That code will initialize the <textarea> with the string from the `GetData(someParameter)` method, which is placed between the opening and closing tag of the <textarea>. This is the recommended practice for initializing a <textarea> in HTML. What I mean by that is, according to https://www.w3schools.com/tags/tag_textarea.asp, a <textarea> **does not**, for example, have/support a `value` attribute like other `<input>` elements.

Then, because we also want to get any changes that the user may make to the text, we wire up the `onchange` event. This should conceptually and effectively give us two-way binding of the <textarea>. In the first example of the following documentation, https://learn.microsoft.com/en-us/aspnet/core/blazor/components/data-binding?view=aspnetcore-6.0, it says the same thing:
> As a demonstration of how data binding composes in HTML, the following example binds the InputValue property to the second <input> element's value and onchange attributes ([change](https://developer.mozilla.org/docs/Web/API/HTMLElement/change_event)). The second <input> element in the following example is a concept demonstration and isn't meant to suggest how you should bind data in Razor components.
image

Again, the reason why we don't simply just use a `@bind` attribute on the <textarea> is because we don't have a bindable property. The data is coming from a method.

Initially, this code appears to work. However, in the following [complete example](https://github.com/Kevin-Andrew/BlazorTextAreaNotUpdating), it does not work as expected. You can download the example from the repo, but I'll also outline and include most of the details and code directly here too for others to quickly see and read.

Create a new solution/project using either the Blazor WebAssembly App or Blazor Server App project templates:
image
In the `Index.razor` page, replace the code with the following:
```
@page "/"

<PageTitle>Index</PageTitle>

<h1>Instructions</h1>
<br/>
<ol>
    <li>Click one of the links below. A textarea and text box will be rendered with some initial text.</li>
    <li>Click back and forth between each link. Notice that both textarea and text box will be populated
        with, and change between, the initial/default values assigned to each link.
    </li>
    <li>Click the Method 1 link, then change the value of the text in the <b>textarea</b>.</li>
    <li>Then click the Method 2 link. Notice that the text in the textarea did not get updated/changed
        to the Method 2 text, as the text box did. This doesn't seem like correct behavior.
    </li>
</ol>

<a href="./?method=method1">Method 1</a>
<br />
<a href="./?method=method2">Method 2</a>

@if (!String.IsNullOrWhiteSpace(Method))
{
    <form>
        @* In HTML, according to https://www.w3schools.com/tags/tag_textarea.asp, the text that is to be
        displayed in the textarea should be placed between the opening and closing <textarea> tags. *@
        <textarea rows="3" style="width: 350px;" @onchange="Text_OnChange">@GetData(Method)</textarea>
        <br />
        <input type="text" style="width: 350px;" value="@GetData(Method)" @onchange="Text_OnChange" />
        <br/>
        <br/>
        @* If, however, we make the <textarea> similar to the <input type="text"> by specifying a 'value'
        attribute, then the behavior matches that of the <input type="text">, as desired. However, according
        to https://www.w3schools.com/tags/tag_textarea.asp, the <textarea> doesn't have/support a 'value'
        attribute.
        (Also, while this workaround happens to work in this simple sample, I have another project where
        this workaround doesn't fix the problem.) *@
        @* Uncomment the following code to text this workaround. *@
        @*<textarea rows="3" style="width: 350px;" value="@GetData(Method)" @onchange="Text_OnChange"></textarea>*@
    </form>
}

@code
{
    private Dictionary<string, string> cache = new Dictionary<string, string>
    {
        { "method1", "Initial value for method 1" },
        { "method2", "Method 2 initial value." },
    };

    [Parameter, SupplyParameterFromQuery(Name = "method")]
    public string? Method { get; set; }

    protected override Task OnInitializedAsync()
    {
        return base.OnInitializedAsync();
    }

    protected override void OnParametersSet()
    {
        base.OnParametersSet();
    }

    private void Text_OnChange(ChangeEventArgs e)
    {
        Console.WriteLine($"Text_OnChange: {e.Value}");

        if (Method != null && cache.ContainsKey(Method))
        {
            cache[Method] = (string)e.Value!;
        }
    }

    private string GetData(string key)
    {
        Console.WriteLine($"Data for {key} is: {cache[key]}");
        return cache[key];
    }
}
```
Run/debug the app and follow the instructions by clicking on both links to see the initial data:
image
and
image

Then change the text in the <textarea> of the first link:
image

Finally, click on the "Method 2" link and notice that the text in the <textarea> is not updated:
image

As you see in the example, I have also include an `<input type="text">` element. With an `<input type="text">` element, the proper way to set/initialize the text value is to set the `value` attribute. So we can accomplish what we need to do, that is, conceptually/effectively perform two-way binding by setting the `value` attribute to the result of the `GetData(someParameter)` method and wiring up and handling the `@onchange` event.

If you noticed in the last screen shot, the text in the regular text  box was correctly updated. So why wasn't the <textarea>? We can take it a step further and have the user change the text in the regular text box too, then click back and forth between the two links and the text in the text box is always updated correctly:
image

So why is the behavior between those to, very similar elements, different?

Doing more tests on my own, I also discovered that if you change the <textarea> to instead use/have a `value` attribute, which again, according to the HTML specification is not a valid attribute, then the behavior is as expected and matches that of the regular `<input type="text">` element. At least, in this simple example, that workaround yields the correct/desired behavior. However, in my real project, that workaround still doesn't give the correct/expected behavior.

I have found and read the following:
https://github.com/dotnet/aspnetcore/issues/45797
I have also read the following, that is new in .NET 7.
https://learn.microsoft.com/en-us/aspnet/core/blazor/components/data-binding?view=aspnetcore-7.0#use-bindgetbindset-modifiers-and-avoid-event-handlers-for-two-way-data-binding
But in my example, I'm not trying to do. Specifically, I am not attempting to modify the value that the user enters and expect Blazor to render that modified value.

I have not tried the new `@bind:get/@bind:set` modifiers available in .NET 7 to know whether or not it would fix my problem. But if the new `@bind:get` only supports binding to a C# property, then no, that won't solve my problem because I need to pass in a parameter to a method.

### Expected Behavior

Having said all of that, I have worked around all of the issues, but I wanted to report this nonetheless and at least try to get an answer to the following:
Why, in the provided example code, is there a behavior difference between a <textarea> that is written as:
```
<textarea rows="3" style="width: 350px;" @onchange="Text_OnChange">@GetData(Method)</textarea>
```
versus being written as:
```
<textarea rows="3" style="width: 350px;" value="@GetData(Method)" @onchange="Text_OnChange"></textarea>
```

To me, the second one, that uses the `value` attribute, shouldn't even be allowed. But at a minimum, isn't it a bug that those two behave differently?

### Steps To Reproduce

https://github.com/Kevin-Andrew/BlazorTextAreaNotUpdating

### Exceptions (if any)

_No response_

### .NET Version

.NET 6

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by NanthiniMahalingam**
### Description
When a <textarea> is initialized with content placed between its opening and closing tags (as per the HTML spec), and an onchange handler is wired up for two-way binding, the textarea's displayed value does not update when the component re-renders with new data — even though a sibling <input type="text"> using a value attribute updates correctly.

### Root cause
When Blazor's interactive renderer (WebAssembly/Server) processes an updateText diff edit for a <textarea>'s child text node, it only sets textNode.textContent. Browsers decouple the value property (the displayed/user-editable content) from textContent/defaultValue once the user has edited the field. So subsequent Blazor re-renders only update the defaultValue, leaving the visible value stale.

### Fix detail
When a <textarea> element's child text node is updated during a Blazor re-render, sync the textarea's value property to keep the displayed content in sync, as browsers decouple the user-editable value from the underlying text content once the field has been interacted with.

### Testing
Added the Unit test for the text area not updating case. After adding the fix, text area value is updated when current page parameter changes.

### Output

Before changes

https://github.com/user-attachments/assets/61c7f28e-6d8b-44ee-8ac7-ab0b7ea74651

After Changes

https://github.com/user-attachments/assets/b023716f-26b3-4d3e-b7ba-4b01169d6ec1

### CODE REVIEW

File: src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/TextareaMethodInteractive.razor [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        { "method2", "Method 2 initial value." },
+    };
+
+    [Parameter, SupplyParameterFromQuery(Name = "method")]
+    public string? Method { get; set; }
+
+    protected override void OnParametersSet()
+    {
+        base.OnParametersSet();
+    }
```
[the remark this answers is not in the document]
Comment (NanthiniMahalingam) [reply src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/TextareaMethodInteractive.razor]: - I have removed the redundant OnParametersSet override in TextareaMethodInteractive.razor page.

File: src/Components/test/E2ETest/Tests/TextareaMethodInteractiveTest.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+using Components.TestServer.RazorComponents;
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure;
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure.ServerFixtures;
+using Microsoft.AspNetCore.E2ETesting;
+using OpenQA.Selenium;
+using TestServer;
+using Xunit.Abstractions;
+
+namespace Microsoft.AspNetCore.Components.E2ETests.Tests;
```
[the remark this answers is not in the document]
Comment (NanthiniMahalingam) [reply src/Components/test/E2ETest/Tests/TextareaMethodInteractiveTest.cs]: - I have updated the namespace to Microsoft.AspNetCore.Components.E2ETest.Tests for consistency.

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts [hunk 3]
```diff
@@ -180,6 +180,13 @@ export class BrowserRenderer {
           const textNode = getLogicalChild(parent, childIndexAtCurrentDepth + siblingIndex);
           if (textNode instanceof Text) {
             textNode.textContent = frameReader.textContent(frame);
+            // If this text node is inside a <textarea>, also update the textarea's 'value' property.
+            // Browsers track the displayed value (the 'value' property) separately from the initial
+            // text content (defaultValue/textContent) once the user has edited the field, so updating
+            // textContent alone won't update what the user sees.
+            if (textNode.parentElement instanceof HTMLTextAreaElement) {
+              textNode.parentElement.value = textNode.parentElement.textContent || '';
+            }
```
[the remark this answers is not in the document]
Comment (NanthiniMahalingam) [reply src/Components/Web.JS/src/Rendering/BrowserRenderer.ts]: - I have stored textNode.parentElement in a local parentElement variable and reuse it instead of accessing the property twice.

**Comment by NanthiniMahalingam** [comment]
@dotnet-policy-service agree company="Syncfusion, Inc."

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts [hunk 4]
```diff
@@ -179,7 +179,17 @@ export class BrowserRenderer {
...
+            const newText = frameReader.textContent(frame);
+            textNode.textContent = newText;
+            // If this text node is inside a <textarea>, also update the textarea's 'value' property.
+            // Browsers track the displayed value (the 'value' property) separately from the initial
+            // text content (defaultValue/textContent) once the user has edited the field, so updating
+            // textContent alone won't update what the user sees. We skip this while the textarea has
+            // focus to avoid clobbering the caret position of a user who is actively typing.
+            const parentElement = textNode.parentElement;
+            if (parentElement instanceof HTMLTextAreaElement && document.activeElement !== parentElement) {
+              parentElement.value = newText || '';
```
Comment (kotlarmilos) [inline src/Components/Web.JS/src/Rendering/BrowserRenderer.ts]: newText contains only the changed text frame, not the complete textarea content. Compiler can emit multiple text frames for markup such as <textarea>Hello ...!</textarea>, so this assignment can replace the value with only the updated segment and discard the rest

Comment (NanthiniMahalingam) [reply src/Components/Web.JS/src/Rendering/BrowserRenderer.ts]: Hi @kotlarmilos ,

- I've now reconstructed the complete textarea value by iterating through all child text nodes and concatenating content from every Text node child while filtering out non-text nodes.
- This ensures no content is lost regardless of how the compiler splits the textarea markup across multiple text frames.

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts [hunk 5]
```diff
@@ -179,7 +179,18 @@ export class BrowserRenderer {
           const siblingIndex = editReader.siblingIndex(edit);
           const textNode = getLogicalChild(parent, childIndexAtCurrentDepth + siblingIndex);
           if (textNode instanceof Text) {
-            textNode.textContent = frameReader.textContent(frame);
+            const newText = frameReader.textContent(frame);
+            textNode.textContent = newText;
+            const parentElement = textNode.parentElement;
+            if (parentElement instanceof HTMLTextAreaElement && document.activeElement !== parentElement) {
+              let fullContent = '';
+              for (const node of Array.from(parentElement.childNodes)) {
```
Comment (kotlarmilos) [inline src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:189]: An updateText edit means the rendered value actually changed. If the textarea is focused, this code updates only textContent, leaving the displayed .value stale. Can we do something this instead?
```ts
if (textarea instanceof HTMLTextAreaElement) {
  const value = textarea.textContent ?? '';
  if (textarea.value !== value) {
    textarea.value = value;
  }
}
```

Comment (NanthiniMahalingam) [reply src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:189]: Hi @kotlarmilos ,
I've reconstructed the solution: textarea .value is rebuilt from all text nodes, but only when not focused — this ensures multi-frame textareas sync correctly while preserving the user's caret position during editing. Please let us know if you have concerns.

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts [hunk 6]
```diff
@@ -179,7 +179,21 @@ export class BrowserRenderer {
...
+                if (node instanceof Text) {
+                  fullContent += node.textContent || '';
+                }
+              }
+              const newValue = fullContent || '';
```
Comment (kotlarmilos) [inline src/Components/Web.JS/src/Rendering/BrowserRenderer.ts]: I believe this check is redundant, fullContent will always be non-empty or ''

File: src/Components/test/E2ETest/Tests/TextareaMethodInteractiveTest.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (kotlarmilos) [inline src/Components/test/E2ETest/Tests/TextareaMethodInteractiveTest.cs]: I suggest removing E2E test, I don't think it is necessary.

File: src/Components/Web.JS/test/BrowserRenderer.updateText.test.ts [hunk 8]
```diff
@@ -0,0 +1,173 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (kotlarmilos) [inline src/Components/Web.JS/test/BrowserRenderer.updateText.test.ts]: I suggest reducing tests to a single test case.

Comment (NanthiniMahalingam) [reply src/Components/Web.JS/test/BrowserRenderer.updateText.test.ts]: Hi @kotlarmilos ,
I have consolidated the test cases into a single test function that covers multiple scenarios: multiple text frames, focus state handling, and edge cases with empty/non-text nodes.

File: src/Components/test/E2ETest/Tests/TextareaMethodInteractiveTest.cs (hunk 7 above)
Comment (NanthiniMahalingam) [reply src/Components/test/E2ETest/Tests/TextareaMethodInteractiveTest.cs]: Hi @kotlarmilos ,
I've removed the E2E test case. The unit tests now comprehensively cover the updateText logic including multiple text frames, focus state handling, and edge cases, making the E2E test redundant.

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts (hunk 6 above)
Comment (NanthiniMahalingam) [reply src/Components/Web.JS/src/Rendering/BrowserRenderer.ts]: Hi @kotlarmilos ,
Removed the redundant value check. Since fullContent is always a string (either empty or with content), direct assignment to parentElement.value is safe and simpler than the conditional.

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts [hunk 9]
```diff
@@ -179,7 +179,20 @@ export class BrowserRenderer {
           const siblingIndex = editReader.siblingIndex(edit);
           const textNode = getLogicalChild(parent, childIndexAtCurrentDepth + siblingIndex);
           if (textNode instanceof Text) {
-            textNode.textContent = frameReader.textContent(frame);
+            const newText = frameReader.textContent(frame);
+            textNode.textContent = newText;
+            const parentElement = textNode.parentElement;
+            if (parentElement instanceof HTMLTextAreaElement && document.activeElement !== parentElement) {
```
Comment (kotlarmilos) [inline src/Components/Web.JS/src/Rendering/BrowserRenderer.ts]: Please remove the document.activeElement !== parentElement condition. An updateText edit means the rendered value changed, so skipping the update while the textarea is focused leaves its displayed value stale. The parentElement.value !== fullContent check already avoids unnecessary assignments.

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts [hunk 10]
```diff
@@ -179,7 +179,20 @@ export class BrowserRenderer {
...
+              let fullContent = '';
+              for (const node of parentElement.childNodes) {
```
Comment (kotlarmilos) [inline src/Components/Web.JS/src/Rendering/BrowserRenderer.ts]: This does not compile, look at the build errors

File: src/Components/Web.JS/test/BrowserRenderer.updateText.test.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
+import { expect, test, describe, beforeEach, afterEach } from '@jest/globals';
+import { BrowserRenderer } from '../src/Rendering/BrowserRenderer';
+import { RenderBatch, ArrayBuilderSegment, RenderTreeEdit, RenderTreeFrame, EditType, FrameType, ArrayValues } from '../src/Rendering/RenderBatch/RenderBatch';
+
+describe('BrowserRenderer.updateText with textarea containing multiple text frames', () => {
```
Comment (kotlarmilos) [inline src/Components/Web.JS/test/BrowserRenderer.updateText.test.ts]: This test does not exercise the production change. It creates a BrowserRenderer but never invokes updateComponent and reimplements the synchronization logic in the test

Comment (NanthiniMahalingam) [reply src/Components/Web.JS/test/BrowserRenderer.updateText.test.ts]: Hi @kotlarmilos ,
Based on your suggestion, I have updated the test case and reimplemented the synchronization logic in the test.

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts (hunk 9 above)
Comment (NanthiniMahalingam) [reply src/Components/Web.JS/src/Rendering/BrowserRenderer.ts]: Hi @kotlarmilos ,
Removed the activeElement check to ensure the textarea value is always updated when content changes; the fullContent comparison already prevents unnecessary assignments.

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts (hunk 10 above)
Comment (NanthiniMahalingam) [reply src/Components/Web.JS/src/Rendering/BrowserRenderer.ts]: Hi @kotlarmilos,
Now, I've resolved the build errors.

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts [hunk 12]
```diff
@@ -179,7 +179,20 @@ export class BrowserRenderer {
...
+            const parentElement = textNode.parentElement;
+            if (parentElement instanceof HTMLTextAreaElement) {
+              let fullContent = '';
+              for (const node of Array.from(parentElement.childNodes)) {
+                if (node instanceof Text) {
+                  fullContent += node.textContent || '';
+                }
+              }
+              if (parentElement.value !== fullContent) {
+                parentElement.value = fullContent;
```
Comment (kotlarmilos) [inline src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:195]: This overwrites existing value frame. For  <textarea value="currentValue">defaultValue</textarea> , the initial displayed value comes from currentValue, but changing defaultValue now replaces it with the child content. This also diverges from the static renderer, which defines the value frame as overriding textarea text content. Please do this synchronization only when the textarea value comes from its child content

Comment (NanthiniMahalingam) [reply src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:195]: Hi @kotlarmilos ,

- I've fixed the textarea so it now syncs its value from child content only when there's no explicit value frame (checked via hasDeferredValue), so <textarea value="currentValue">defaultValue</textarea> keeps currentValue and matches the static renderer.
- I also converted should sync textarea value from child content when there is no value frame from a test that reimplemented the sync logic inline (so it passed even without the fix) into a real regression test that drives the actual BrowserRenderer and simulates a user edit (textarea.value = 'user typed') to defeat jsdom's auto-reflect — so it now fails without the fix and passes with it.

File: src/Components/Web.JS/src/Rendering/DomSpecialPropertyUtil.ts [hunk 13]
```diff
@@ -21,6 +21,12 @@ export function tryApplySpecialProperty(element: Element, name: string, value: s
   }
 }
 
+export function hasDeferredValue(element: Element): boolean {
```
Comment (kotlarmilos) [inline src/Components/Web.JS/src/Rendering/DomSpecialPropertyUtil.ts]: Can we inline this function?

**Review by kotlarmilos — APPROVED** [review]
One suggestion, otherwise LGTM!

File: src/Components/Web.JS/src/Rendering/DomSpecialPropertyUtil.ts (hunk 13 above)
Comment (NanthiniMahalingam) [reply src/Components/Web.JS/src/Rendering/DomSpecialPropertyUtil.ts]: > Can we inline this function?

Hi @kotlarmilos ,
I have addressed by inlining the deferred-value check and removing the helper. The shared property name remains centralized to avoid duplicating the internal string.

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts [hunk 14]
```diff
@@ -179,7 +179,22 @@ export class BrowserRenderer {
...
+            // Only synchronize the textarea's value from its child content when there's no explicit
+            // 'value' frame. Otherwise the value frame takes precedence, matching the static renderer.
+            if (parentElement instanceof HTMLTextAreaElement && !(deferredValuePropname in parentElement)) {
+              let fullContent = '';
+              for (const node of Array.from(parentElement.childNodes)) {
+                if (node instanceof Text) {
+                  fullContent += node.textContent || '';
+                }
+              }
+              if (parentElement.value !== fullContent) {
```
Comment (PureWeen) [inline src/Components/Web.JS/src/Rendering/BrowserRenderer.ts]: This synchronizes from an intermediate DOM state if the same render batch later structurally edits another textarea child. `RenderTreeDiffBuilder` can emit `UpdateText` followed by `RemoveFrame`/`PrependFrame` (for example, `RecognizesEquivalentRegions`). With a dirty textarea containing text nodes `A` and `B`, a batch that updates `A` to `A2` and then removes `B` assigns `.value = "A2B"` here; `removeFrame` then leaves `textContent === "A2"`, but nothing resynchronizes `.value`, so the displayed value remains stale as `A2B`. The new tests only cover an isolated `updateText`. Could we collect affected textareas and synchronize them after the edit loop, with a mixed-edit regression test?

This review comment was generated by GitHub Copilot with the help of a human.

Comment (NanthiniMahalingam) [reply src/Components/Web.JS/src/Rendering/BrowserRenderer.ts]: > This synchronizes from an intermediate DOM state if the same render batch later structurally edits another textarea child. `RenderTreeDiffBuilder` can emit `UpdateText` followed by `RemoveFrame`/`PrependFrame` (for example, `RecognizesEquivalentRegions`). With a dirty textarea containing text nodes `A` and `B`, a batch that updates `A` to `A2` and then removes `B` assigns `.value = "A2B"` here; `removeFrame` then leaves `textContent === "A2"`, but nothing resynchronizes `.value`, so the displayed value remains stale as `A2B`. The new tests only cover an isolated `updateText`. Could we collect affected textareas and synchronize them after the edit loop, with a mixed-edit regression test?
> 
> This review comment was generated by GitHub Copilot with the help of a human.

Hi @PureWeen ,

- I have addressed the intermediate DOM synchronization issue by collecting affected textareas during edit processing and synchronizing their values after the complete edit loop. 
- I also added regression coverage for mixed and structural child edits. All Web.JS tests pass.

**Review by kotlarmilos — APPROVED** [review]
LGTM!
